### ISSUE TITLE
YAML dump errors

### ISSUE THREAD

**Issue by ollieeknight** [caleblareau/mgatk#84]
**Describe the bug**

```
mgatk tenx -i DOGMA_sample1_exp1_libA/outs/possorted_bam.bam -n sample -o DOGMA_sample1_exp1_libA/mgatk/ -c 8 -bt CB -b DOGMA_sample1_exp1_libA/outs/filtered_peak_bc_matrix/barcodes.tsv --skip-R
Wed Jan 10 11:24:15 CET 2024: mgatk v0.6.9
Wed Jan 10 11:24:15 CET 2024: Found bam file: DOGMA_sample1_exp1_libA/outs/possorted_bam.bam for genotyping.
Wed Jan 10 11:24:15 CET 2024: Found file of barcodes to be parsed: DOGMA_HSPC_exp1_libA/outs/filtered_peak_bc_matrix/barcodes.tsv
Wed Jan 10 11:24:15 CET 2024: User specified mitochondrial genome matches .bam file
Wed Jan 10 11:26:16 CET 2024: Finished determining/splitting barcodes for genotyping.
Wed Jan 10 11:26:16 CET 2024: Genotyping samples with 8 threads
Traceback (most recent call last):
  File "/fast/scratch/users/knighto_c/ngs/samples/venv3/bin/mgatk", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/fast/scratch/users/knighto_c/ngs/samples/venv3/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/fast/scratch/users/knighto_c/ngs/samples/venv3/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/fast/scratch/users/knighto_c/ngs/samples/venv3/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/fast/scratch/users/knighto_c/ngs/samples/venv3/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/fast/scratch/users/knighto_c/ngs/samples/venv3/lib/python3.11/site-packages/mgatk/cli.py", line 365, in main
    yaml.dump(dict1, yaml_file, default_flow_style=False, Dumper=yaml.RoundTripDumper)
  File "/fast/scratch/users/knighto_c/ngs/samples/venv3/lib/python3.11/site-packages/ruamel/yaml/main.py", line 1251, in dump
    error_deprecation('dump', 'dump', arg="typ='unsafe', pure=True")
  File "/fast/scratch/users/knighto_c/ngs/samples/venv3/lib/python3.11/site-packages/ruamel/yaml/main.py", line 1039, in error_deprecation
    raise AttributeError(s, name=None)
AttributeError:
"dump()" has been removed, use

  yaml = YAML(typ='unsafe', pure=True)
  yaml.dump(...)

instead of file "/fast/scratch/users/knighto_c/ngs/samples/venv3/lib/python3.11/site-packages/mgatk/cli.py", line 365

                        yaml.dump(dict1, yaml_file, default_flow_style=False, Dumper=yaml.RoundTripDumper)
```

**A summary of .log files**

There aren't any :(

**Post an ls -lRh of mgatk_output_folder**

```
(venv3) (base) [knighto_c@hpc-cpu-176 mgatk]$ ls -lRh
.:
total 2.5K
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:28 fasta
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:28 final
drwxr-xr-x 4 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 logs
drwxr-xr-x 4 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 qc
drwxr-xr-x 8 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 temp

./fasta:
total 257K
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 17K Jan 10 11:28 chrM.fasta
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  19 Jan 10 11:28 chrM.fasta.fai

./final:
total 256K
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 119K Jan 10 11:28 chrM_refAllele.txt

./logs:
total 1.5K
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  123 Jan 10 11:30 base.mgatk.log
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 filterlogs
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 rmdupslogs

./logs/filterlogs:
total 0

./logs/rmdupslogs:
total 0

./qc:
total 1.0K
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 depth
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 quality

./qc/depth:
total 0

./qc/quality:
total 0

./temp:
total 3.0K
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 barcoded_bams
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:28 barcode_files
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 quality
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 ready_bam
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 sparse_matrices
drwxr-xr-x 2 knighto_c hpc-ag-romagnani 4.0K Jan 10 11:30 temp_bam

./temp/barcoded_bams:
total 436M
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  58M Jan 10 11:30 barcodes.1.bam
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 2.3K Jan 10 11:30 barcodes.1.bam.bai
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  62M Jan 10 11:30 barcodes.2.bam
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 2.3K Jan 10 11:30 barcodes.2.bam.bai
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  52M Jan 10 11:30 barcodes.3.bam
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 2.3K Jan 10 11:30 barcodes.3.bam.bai
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  46M Jan 10 11:29 barcodes.4.bam
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 2.3K Jan 10 11:30 barcodes.4.bam.bai
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  49M Jan 10 11:30 barcodes.5.bam
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 2.3K Jan 10 11:30 barcodes.5.bam.bai
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  63M Jan 10 11:30 barcodes.6.bam
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 2.3K Jan 10 11:30 barcodes.6.bam.bai
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  54M Jan 10 11:30 barcodes.7.bam
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 2.4K Jan 10 11:30 barcodes.7.bam.bai
-rw-r--r-- 1 knighto_c hpc-ag-romagnani  55M Jan 10 11:30 barcodes.8.bam
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 2.3K Jan 10 11:30 barcodes.8.bam.bai

./temp/barcode_files:
total 2.0M
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 7.1K Jan 10 11:28 barcodes.1.txt
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 7.1K Jan 10 11:28 barcodes.2.txt
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 7.1K Jan 10 11:28 barcodes.3.txt
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 7.1K Jan 10 11:28 barcodes.4.txt
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 7.1K Jan 10 11:28 barcodes.5.txt
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 7.1K Jan 10 11:28 barcodes.6.txt
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 7.1K Jan 10 11:28 barcodes.7.txt
-rw-r--r-- 1 knighto_c hpc-ag-romagnani 7.1K Jan 10 11:28 barcodes.8.txt

./temp/quality:
total 0

./temp/ready_bam:
total 0

./temp/sparse_matrices:
total 0

./temp/temp_bam:
total 0
```

**Describe the sequencing assay being analyzed**

DOGMA-seq. It is run in ATAC only mode as we have had some inefficient runs and I want to recover as many cells for each modality (GEx, ATAC), separately. However, I've also tested this on ASAP-seq data and it won't work either

**Clarify if the execution successful on the test data provided in the repository**

I tried it, but it also failed with the exact same log and errors.

**Additional context**

Thanks so much for the tool, and all of the help you provide me!

**Comment by ollieeknight** [comment]
I've gotten the test to work, by specifying some installation versions:

```
conda create -y -n mitochondrial_genotyping openjdk r-base=4.2.3 r-data.table r-matrix bioconductor-genomicranges bioconductor-summarizedexperiment ruamel.yaml==0.17.35 snakemake==7.32.4 pyyaml==6.0.1 yaml=0.2.5
```

**Comment by TaichiHIBI** [comment]
I'm facing the exact same problem.

**Comment by TaichiHIBI** [comment]
Adding the following script according to the instructions produced this additional error.

I added at the beginning
from ruamel import yaml
from ruamel.yaml.main import YAML

Added on lines 366,367,412,413
yaml = YAML(typ='rt')
yaml.preserve_quotes = True

Output
AttributeError: 'YAML' object has no attribute 'RoundTripDumper'

**Comment by ollieeknight** [comment]
so, creating my conda env above, activating and then using `pip install mgatk` should work perfectly, I just ran `mgatk` for 8 or so samples and it worked perfectly. I think it's to do with the dump() function being depracated, [here](https://pypi.org/project/ruamel.yaml/0.18.5/), specifically:

"As announced, in 0.18.0, the old PyYAML functions have been deprecated. (`scan`, `parse`, `compose`, `load`, `emit`, `serialize`, `dump` and their variants (`_all`, `safe_`, `round_trip_`, etc))."

So, I suppose @caleblareau it should be necessary to either update the installation requirements, or update the `cli.py` for the latest `ruamel.yaml` package. I am happy to do a PR for the first but I don't have enough experience to do the second!

**Comment by caleblareau** [comment]
```
that also means you, or the person developing your program, has not tested with warnings on
```

oof, right where it hurts :) 

@ollieeknight @TaichiHIBI I just posted version 0.7.0 with the yaml update to `pip`. Try upgrading and let me know if the error persists?

**Comment by ollieeknight** [comment]
ha! I saw that error message, but I didn't want to include it felt a bit too accusatory... Works perfect, thanks.

Unfortunately I'm now getting this issue with one of my samples - it works for two others but fails here. Any recomendations?

```R
>   mtDNA_data <- ReadMGATK('multiome_IBD_sample/mgatk/final/')
Reading allele counts
Reading metadata
Building matrices
Error in fixupDN.if.valid(value, x@Dim) : 
  length of Dimnames[[2]] (20162) is not equal to Dim[2] (20161)
```

**Comment by caleblareau** [comment]
great! glad it worked.

is this from the pipeline or you running signac? 

are there RDS files in the output folder of that sample?

**Comment by ollieeknight** [comment]
yep, both the seurat and signac files. For a sample which I know worked for `ReadMGATK` (with a good scatter distribution versus nCount_ATAC) I also imported the `signac.rds` file via `readRDS`. Shouldn't the correlation between the two be 1 = 1? Perhaps I misunderstand what the 'depth' column in the `signac.rds` file is.

![image](https://github.com/caleblareau/mgatk/assets/85193902/78b1f4ab-bb70-4cff-9bcf-1cbb4df588d9)

**Comment by ollieeknight** [comment]
yep, all good if i read it through. not sure why it didnt work with `ReadMGATK`!

**Comment by ollieeknight** [comment]
if anyone is running into any errors atm with [this](https://github.com/snakemake/snakemake/issues/2607), a fix is to make sure `pulp=2.7.0` is installed, for some reason 2.8.0 breaks snakemake at the moment
